### PR TITLE
lsns: print namespace tree based on the relationship (parent/child or owner/owned)

### PR DESCRIPTION
**Pull request by masatake**
Introduce new option -T[parent|owner]|--nstree[=parent|=owner].

With this change, lsns prints parent/child relationship tree if
"-T parent" is given and owner/owned relationship tree if "-T owner is given.

Passing only "-T" is the same as "-Towner."

Example sessions:

    # ./lsns -Tparent -ons,type,pns| head -20
    NS             TYPE          PNS
    4026531837     user            0
    ├─4026532508   user   4026531837
    ├─4026532609   user   4026531837
    ├─4026532610   user   4026531837
    ├─4026532629   user   4026531837
    ├─4026532705   user   4026531837
    ├─4026532901   user   4026531837
    ├─4026533090   user   4026531837
    ├─4026533185   user   4026531837
    ├─4026533280   user   4026531837
    └─4026533468   user   4026531837
    4026531835     cgroup          0
    4026531836     pid             0
    └─4026533038   pid    4026531836
      ├─4026532934 pid    4026533038
      ├─4026533715 pid    4026533038
      ├─4026533716 pid    4026533038
      ...

    # ./lsns -Towner -ons,type,ons| head -20
    NS             TYPE          ONS
    4026531837     user            0
    ├─4026531835   cgroup 4026531837
    ├─4026531836   pid    4026531837
    ├─4026531838   uts    4026531837
    ├─4026531839   ipc    4026531837
    ├─4026531840   mnt    4026531837
    ├─4026531861   mnt    4026531837
    ├─4026532001   net    4026531837
    ├─4026532219   mnt    4026531837
    ├─4026532357   mnt    4026531837
    ├─4026532383   net    4026531837
    ├─4026532475   mnt    4026531837
    ├─4026532476   mnt    4026531837
    ├─4026532504   mnt    4026531837
    ├─4026532508   user   4026531837
    │ ├─4026532509 ipc    4026532508
    │ └─4026532511 net    4026532508
    ├─4026532573   mnt    4026531837
    ├─4026532574   mnt    4026531837

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

File: sys-utils/lsns.8.adoc [hunk 1]
```diff
@@ -63,6 +63,12 @@ Do not truncate text in columns.
 *-W*, *--nowrap*::
 Do not use multi-line text in columns.
 
+*-T*, *--nstree* _rel_::
```
Comment (karelzak) [inline sys-utils/lsns.8.adoc:66]: Cannot we use --tree rather than --nstree?

Comment (karelzak) [reply sys-utils/lsns.8.adoc:66]: but I see we already use the tree-like output for processes in namespace (lsns <ns>) too. Maybe we can use

 --tree[=parent|owner|process]

and make --tree=process default for `lsns [<ns>]`, and --tree=owner default for `lsns --tree`. What do you think?

File: sys-utils/lsns.c [hunk 2]
```diff
@@ -1080,6 +1150,15 @@ int main(int argc, char *argv[])
 		case 'W':
 			ls.no_wrap = 1;
 			break;
+		case 'T':
+			ls.nstree = LSNS_NSTREE_OWNER;
+			if (optarg) {
```
Comment (karelzak) [inline sys-utils/lsns.c:1155]: You need:
```C
if (*optarg == '=')
  optarg++;
```
to make it robust, getopt_long() is horrible bastard when use optional arguments :-)

**Comment by karelzak** [comment]
Looks good.

File: sys-utils/lsns.8.adoc (hunk 1 above)
Comment (masatake) [reply sys-utils/lsns.8.adoc:66]: I think what you wrote is better. Updated my change.

File: sys-utils/lsns.c (hunk 2 above)
Comment (masatake) [reply sys-utils/lsns.c:1158]: Updated my change. Thank you.

**Comment by masatake** [comment]
Should I squash the commits into one?

**Comment by karelzak** [comment]
> Should I squash the commits into one?

No, I think incremental changes are fine.

**Comment by masatake** [comment]
Do I have items to do for merging?

**Comment by karelzak** [comment]
> Do I have items to do for merging?

I'll merge into the "next" branch and to "master" after v2.37 release. Now it's too late for 2.37.

File: sys-utils/lsns.c [hunk 3]
```diff
@@ -890,6 +938,28 @@ static struct libscols_table *init_scols_table(struct lsns *ls)
...
+static void show_namespace(struct lsns *ls, struct libscols_table *tab,
+			   struct lsns_namespace *ns, struct lsns_process *proc)
+{
+	/*
+	 * create a tree from owner->owned and/or parent->child relation
+	 */
+	if (ls->tree == LSNS_TREE_OWNER
+	    && ns->ownerns
+	    && !ns->ownerns->ns_outline)
+		show_namespace(ls, tab, ns->ownerns, proc);
```
Comment (masatake) [inline sys-utils/lsns.c:950]: passing proc may be wrong. Instread ns->owners->proc should be passed.

Comment (masatake) [reply sys-utils/lsns.c]: Pushed one commit for fixing this.

**Comment by karelzak** [comment]
Merged to the "next" branch and it will be merged to the "master" later after v2.37 release. Thanks!

**Comment by karelzak** [comment]
Merged to the master branch. Thanks!

**Pull request closed without merging**